source/common/uma_helpers.hpp

```diff
@@ -0,0 +1,142 @@
+        if constexpr (std::is_same_v<decltype(function_return_type(ObjMember)), void>) {
+            try {
+                (reinterpret_cast<ObjT *>(obj)->*ObjMember)(args...);
+            } catch (...) {
+                // TODO: log
+            }
+        } else if constexpr (std::is_same_v<decltype(function_return_type(ObjMember)), enum uma_result_t>) {
+            try {
+                return (reinterpret_cast<ObjT *>(obj)->*ObjMember)(args...);
+            } catch (...) {
+                return UMA_RESULT_ERROR_UNKNOWN;
+            }
+        } else {
+            try {
+                return (reinterpret_cast<ObjT *>(obj)->*ObjMember)(args...);
+            } catch (...) {
+                return decltype(function_return_type(ObjMember)){};
+            }
+        }
+    };
+}
+
+#define UMA_OPS_GENERATE_FUNCTION(ObjT, member, ops) umaOpsGenerateFunction<ObjT, &decltype(ops)::member, &ObjT::member>(ops)
+
+template <typename ObjT, typename OpsT, typename CtorT, typename DtorT, typename... Args>
+auto createUmaObject(OpsT &ops, CtorT &&ctor, DtorT &&dtor, Args &&...args) {
+    auto argsTuple = std::make_tuple(std::forward<Args>(args)..., (void *)nullptr);
+
+    ops.version = UMA_VERSION_CURRENT;
+    ops.initialize = [](void *params, void **obj) {
+        try {
+            auto *tuple = reinterpret_cast<decltype(argsTuple) *>(params);
+            std::get<std::tuple_size_v<decltype(argsTuple)> - 1>(*tuple) = obj;
+            *obj = createFromTuple<ObjT>(std::move(*tuple));
+            return UMA_RESULT_SUCCESS;
+        } catch (...) {
+            return UMA_RESULT_ERROR_UNKNOWN;
+        }
+    };
+    ops.finalize = [](void *obj) {
+        delete reinterpret_cast<ObjT *>(obj);
+    };
+    return makeUmaHandle(ctor, dtor, &ops, reinterpret_cast<void *>(&argsTuple));
+}
+
+/// @brief creates UMA memory provider based on given MemProvider type.
+/// MemProvider should implement all functions defined by
+/// uma_memory_provider_ops_t, except for intialize and finalize (those are
+/// replaced by ctor and dtor). All arguments passed to this function are
+/// forwarded to MemProvider ctor.
+template <typename MemProvider, typename... Args>
+auto memoryProviderMakeUnique(Args &&...args) {
+    uma_memory_provider_ops_t ops;
+
+    UMA_OPS_GENERATE_FUNCTION(MemProvider, alloc, ops);
+    UMA_OPS_GENERATE_FUNCTION(MemProvider, free, ops);
+    UMA_OPS_GENERATE_FUNCTION(MemProvider, get_last_result, ops);
+
+    return createUmaObject<MemProvider>(ops, &umaMemoryProviderCreate, &umaMemoryProviderDestroy, args...);
+}
+
+/// @brief creates UMA memory pool based on given MemPool type.
+/// MemPool should implement all functions defined by
+/// uma_memory_pool_ops_t, except for intialize and finalize (those are
+/// replaced by ctor and dtor). All arguments passed to this function are
+/// forwarded to MemProvider ctor.
+template <typename MemPool, typename... Args>
+auto poolMakeUnique(Args &&...args) {
```

Comment:
Instead of doing all this, wouldn't it be simpler and more type safe to have some abstract class that implements a conversion operator to uma types?

Reply:
But you still need a way to create a uma handle from that object (call umaCreatePool/Provider).

Reply:
I've removed this file (and tests) from this PR for now. I'll add them once we agree on API and https://github.com/oneapi-src/unified-runtime/pull/322 is merged